MSFragger and MS-GF+ in Skyline; downloads as-needed

chambm opened the pull request:
* added SearchEngine and Ms2Analyzer options on SearchSettingsControl, with new search engines options MSFragger and MS-GF+
* added on-the-fly downloading of MSFragger and MS-GF+ in SearchSettingsControl; MSFragger gets a special dialog for submitting name/email/institution just like FragPipe does
* added Formulas for all isotopic labels that have a defined number of carbon/nitrogen atoms (i.e. they're specific to a single AA)
* added FullNames dictionary in AminoAcidFormulas
* improved stderr capture for ProcessRunner
* fixed ImportPeptideSearchDlg height for DDA search
* fixed MSAmanda not respecting mod terminal specificity for mods that map to its unimod dictionary (this should change MSAmanda results only if terminal mods were used)
* fixed BiblioSpec pepXML reader to use a tolerance to match existing mods due to some weirdness in Crux Percolator's pepXML mod masses

TODO:
- [x] Regenerate all affected audit logs.
- [x] Fix TestModificationMatcher failure.
- [ ] Add infrequent automated testing of original download URLs (the current code has functional tests always redirect to S3); this task may be appropriate for a later PR
- [ ] Run files through Crux Percolator in a single invocation; definitely in a later PR and it probably needs fixes in Crux Percolator to work

chambm commented on pwiz_tools/Skyline/Alerts/MsFraggerDownloadDlg.cs:
```diff
@@ -0,0 +1,185 @@
+        {
+            InitializeComponent();
+
+            const string BOLD_PATTERN = "<b>(.*?)</b>";
```
@brendanx67 As long as the translators preserve the <b> and <link> tags, the highlighting and linking should work in the translated text too.

nickshulman commented on pwiz_tools/Skyline/Util/UtilInstall.cs:
```diff
@@ -114,6 +117,94 @@ public void Dispose()
+                    // TODO: make this conditional somehow so that once per day or week we run with the original URLs to test that they still work
+                    var downloadUrl = requiredFile.DownloadUrl;
+                    if (Program.FunctionalTest)
+                        downloadUrl = new Uri(Regex.Replace(downloadUrl.OriginalString, ".*/(.*)", $"{SKYLINE_TOOL_TESTING_MIRROR_URL}/$1"));
```
The test should be responsible for setting some variable like "TestUrlOverride".
It's not good for this code here to be checking whether "Program.FunctionalTest" is true and then decide where to download from, since it makes it impossible to write a functional test which tests downloading from the real server.

chambm replied:
My current thinking is that in a near-future PR I'll add some Program level property that will be set on TestRunner to enable testing real URLs. There will be a TC config that runs that once per day or per week (haven't decided). Having that logic here seemed awfully convenient. I can't think of a reason to let tests decide on a case by case basis whether to use the test-only URLs or to use the original ones.

nickshulman reviewed (commented):
Looks pretty good. Here are some things that I noticed.

chambm commented on pwiz_tools/Skyline/TestFunctional/DdaSearchTest.cs:
```diff
@@ -18,24 +18,58 @@
 namespace pwiz.SkylineTestFunctional
 {
     [TestClass]
     public class DdaSearchTest : AbstractFunctionalTest
```
I figure they have too much logic in common with each other to bother with inheritance. Only the specific setting names, the results, and the download on-the-fly handling is different.

chambm commented on pwiz_tools/Skyline/TestFunctional/DdaSearchTest.cs:
```diff
@@ -182,9 +286,42 @@ private void TestAmandaSearch()
+                    RunUI(() =>
+                    {
+                        msfraggerDownloaderDlg.SetValues("Matt Chambers (testing download from Skyline)", "matt.chambers42@gmail.com", "UW");
+                        msfraggerDownloaderDlg.ClickAccept();
```
The Accept button starts the download and closes the modal dialog after the download finishes. So yeah, I probably could switch back to using OkDialog. I had switched away from it because I was trying to catch the LongWaitDlg but that turned out to be impossible (and unnecessary) because it's finished when the modal dialog returns.

chambm commented on pwiz_tools/Skyline/Model/DdaSearch/MsgfPlusSearchEngine.cs:
```diff
@@ -110,61 +122,72 @@ public class MsgfPlusSearchEngine : AbstractDdaSearchEngine, IProgressMonitor
-                        @"-Xmx8G -jar MSGFPlus.jar " +
-                        $@"-s {spectrumFilename} -d {FastaFileNames[0]} -tda 1" +
+                    var psi = new ProcessStartInfo(JavaDownloadInfo.JavaBinary,
+                        @"-Xmx16G -jar """ + MsgfPlusBinary + @""" -tasks -2 " +
```
I pity those users. :( But yeah, how about `max(16GB, totalMemory/2)`?

chambm replied:
I meant min, not max.

chambm commented on pwiz_tools/Shared/Common/Chemistry/AminoAcidFormulas.cs:
```diff
@@ -53,7 +53,34 @@ public class AminoAcidFormulas : Immutable
-            {"Val",'V'}
+            {"Val",'V'},
+            {"Sec",'U'},
+            {"Pyl",'O'}
```
They're already there, just out of order. It's somebody else's fault for adding it to DefaultFormas but not LongNames. ;) (Although I wasn't even aware that O and U had 3-letter names until I made this change).

chambm commented on pwiz_tools/Skyline/Alerts/MsFraggerDownloadDlg.cs:
```diff
@@ -0,0 +1,185 @@
+            Download();
+
+            DialogResult = DialogResult.OK;
+            Close();
```
I had forgotten that. Apparently it causes it to Hide rather than Close, but inside a using block it's pretty much the same thing.

chambm commented on pwiz_tools/Skyline/Model/DdaSearch/MsconvertDdaConverter.cs:
```diff
@@ -0,0 +1,175 @@
+                OriginalSpectrumSources = SearchEngine.SpectrumFileNames;
+                ConvertedSpectrumSources = new MsDataFileUri[OriginalSpectrumSources.Length];
+
+                progressMonitor?.UpdateProgress(_progressStatus.ChangeMessage(Resources.MsconvertDdaConverter_Run_Starting_msconvert_conversion_));
```
If progressMonitor is null, then progressStatus.ChangeMessage won't be called, right? It short-circuits the logic AFAIK. But your idea of a Func-based method would look cleaner so I'll switch to that.

nickshulman replied:
Oops! You're right. I forgot about the short circuiting of "?." with function calls.

chambm commented on pwiz_tools/Shared/Common/SystemUtil/ProcessStreamReader.cs:
```diff
@@ -86,11 +91,13 @@ public string ReadLine()
             return ReadLine(null);
         }
 
+        public string ErrorLines => _errorLines?.ToString() ?? string.Empty;
```
Would this be fixed by just `lock (_readLines) return _errorLines?.ToString() ?? string.Empty`? The StringBuilder is only modified inside a `lock (_readLines)`. I'll convert to method as well. I think using lock would be more client-code-friendly than throwing before the process has exited.

chambm commented on pwiz_tools/Skyline/Alerts/MsFraggerDownloadDlg.cs:
```diff
@@ -0,0 +1,185 @@
+
+namespace pwiz.Skyline.Alerts
+{
+    public partial class MsFraggerDownloadDlg : Form
```
That was interesting. Changing it to FormEx broke my RichTextBox bold/underline/color changes. Any idea why? When I moved that code to an OnLoad override it started working again.

nickshulman commented:
I believe the way we got "UnimodData.cs" was that we ran the program in "pwiz_tools\Skyline\Executables\UniModCompiler". If you want to make it so those modifications get Formula's you should probably modify that program.

It might be safer to leave UnimodData.cs along, and instead use "SequenceMassCalc.GetHeavyFormula" when you want to know the formula of the modification. I am a little worried that users might see the message "This modification does not match the Unimod specifications for '{0}'." from EditStaticModDlg if we change UnimodData.cs like this.

I am not an expert on UnimodData.cs so I am not sure what the implications of changing it like this are.

chambm commented:
> I believe the way we got "UnimodData.cs" was that we ran the program in "pwiz_tools\Skyline\Executables\UniModCompiler". If you want to make it so those modifications get Formula's you should probably modify that program.
> 
> It might be safer to leave UnimodData.cs along, and instead use "SequenceMassCalc.GetHeavyFormula" when you want to know the formula of the modification. I am a little worried that users might see the message "This modification does not match the Unimod specifications for '{0}'." from EditStaticModDlg if we change UnimodData.cs like this.
> 
> I am not an expert on UnimodData.cs so I am not sure what the implications of changing it like this are.

That's how I originally got my updated UnimodData.cs. It turns out there's tons of accumulated changes in Unimod that have never been updated in Skyline, and it breaks quite a few tests to update it. Or at least it did when I used my modified compiler which gave formulas and UnimodIds for isotope label mods that were AA-specific, but I know not all the test failures I got were from my compiler changes. But I've already reverted all that jazz and added `Modification.GetAminoAcidIsotopeLabelMassDiff()`. I didn't know about `GetHeavyFormula`. That would have been another way to do it.
